### ISSUE TITLE
[PATCH] simplify Perl_xs_apiversion_bootcheck

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#14215]
Migrated from [rt.perl.org#123146](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=123146) (status was 'resolved')

Searchable as RT123146$

**Comment by p5pRT** [comment]
### From @bulk88

#### Created by @bulk88



See attached patch\.

<details><summary>Perl Info</summary>

```

Flags:
                    category=core
                    severity=low

Site configuration information for perl 5.21.4:

Configured by Owner at Thu Sep 18 12:08:58 2014.

Summary of my perl5 (revision 5 version 21 subversion 4) configuration:
                  Derived from: 7d2b2edb94ab56333b9049a3e26d15ea18445512
                  Ancestor: 19be3be6968e2337bcdfe480693fff795ecd1304
                  Platform:
                    osname=MSWin32, osvers=5.1,
archname=MSWin32-x86-multi-thread
                    uname=''
                    config_args='undef'
                    hint=recommended, useposix=true, d_sigaction=undef
                    useithreads=define, usemultiplicity=define
                    use64bitint=undef, use64bitall=undef, 
uselongdouble=undef
                    usemymalloc=n, bincompat5005=undef
                  Compiler:
                    cc='cl', ccflags ='-nologo -GF -W3 -O1 -MD -Zi -DNDEBUG
-DWIN32
-D_CONSOLE -DNO_STRICT  -DPERL_TEXTMODE_SCRIPTS
-DPERL_HASH_FUNC_ONE_AT_A_TIME -DPERL_IMPLICIT_CONTEXT
-DPERL_IMPLICIT_SYS -DUSE_PERLIO -D_USE_32BIT_TIME_T',
                    optimize='-O1 -MD -Zi -DNDEBUG',
                    cppflags='-DWIN32'
                    ccversion='12.00.8168', gccversion='', gccosandvers=''
                    intsize=4, longsize=4, ptrsize=4, doublesize=8,
byteorder=1234
                    d_longlong=undef, longlongsize=8, d_longdbl=define,
longdblsize=8,
longdblkind=0
                    ivtype='long', ivsize=4, nvtype='double', nvsize=8,
Off_t='__int64',
lseeksize=8
                    alignbytes=8, prototype=define
                  Linker and Libraries:
                    ld='link', ldflags ='-nologo -nodefaultlib -debug
-opt:ref,icf
-libpath:"c:\perl521\lib\CORE"  -machine:x86'
                    libpth=C:\PROGRA~1\MIAF9D~1\VC98\lib
                    libs=oldnames.lib kernel32.lib user32.lib gdi32.lib
winspool.lib
comdlg32.lib advapi32.lib shell32.lib ole32.lib oleaut32.lib
netapi32.lib uuid.lib ws2_32.lib mpr.lib winmm.lib  version.lib
odbc32.lib odbccp32.lib comctl32.lib msvcrt.lib
                    perllibs=oldnames.lib kernel32.lib user32.lib gdi32.lib
winspool.lib  comdlg32.lib advapi32.lib shell32.lib ole32.lib
oleaut32.lib  netapi32.lib uuid.lib ws2_32.lib mpr.lib winmm.lib
version.lib odbc32.lib odbccp32.lib comctl32.lib msvcrt.lib
                    libc=msvcrt.lib, so=dll, useshrplib=true,
libperl=perl521.lib
                    gnulibc_version=''
                  Dynamic Linking:
                    dlsrc=dl_win32.xs, dlext=dll, d_dlsymun=undef,
ccdlflags=' '
                    cccdlflags=' ', lddlflags='-dll -nologo -nodefaultlib
-debug
-opt:ref,icf  -libpath:"c:\perl521\lib\CORE"  -machine:x86'

Locally applied patches:
                    uncommitted-changes
                    a0fe7a7e75de29e59f1da0d6822dc06e5be658fe
                    a261faffee83d0145642ab5d1d046c9f813bc497
                    6506ab86ad1602a9ca720fcd30446dce1461d23d
                    7d2b2edb94ab56333b9049a3e26d15ea18445512


@INC for perl 5.21.4:
                    lib
                    C:/perl521/srcnew/lib
                    .


Environment for perl 5.21.4:
                    HOME (unset)
                    LANG (unset)
                    LANGUAGE (unset)
                    LD_LIBRARY_PATH (unset)
                    LOGDIR (unset)
                    PATH=
                    PERL_BADLANG (unset)
                    PERL_JSON_BACKEND=Cpanel::JSON::XS
                    PERL_YAML_BACKEND=YAML
                    SHELL (unset)



















```
</details>

**Comment by p5pRT** [comment]
### From @bulk88

<details><summary>0001-simplify-Perl_xs_apiversion_bootcheck.patch</summary>

```diff
From 05d2351a71c534de4740b7553c12e3eec7870d98 Mon Sep 17 00:00:00 2001
From: Daniel Dragan <bulk88@hotmail.com>
Date: Thu, 6 Nov 2014 23:57:14 -0500
Subject: [PATCH] simplify Perl_xs_apiversion_bootcheck

We control both strings. Perl API versions are not old decimal or alphas
versions. Maints dont increase Perl API ver. Just do a memcmp. Faster and
less machine code. Before 0xA6 bytes of machine code on VC 2003 32b,
after 0x35. This patch is related to [perl #123136].
---
 XSUB.h           |    2 +-
 embed.fnc        |    2 +-
 pod/perlclib.pod |    3 +++
 pod/perldiag.pod |    5 +++++
 proto.h          |    6 +++---
 util.c           |   30 ++++++------------------------
 6 files changed, 19 insertions(+), 29 deletions(-)

diff --git a/XSUB.h b/XSUB.h
index d0fb253..004a0d6 100644
--- a/XSUB.h
+++ b/XSUB.h
@@ -331,7 +331,7 @@ Rethrows a previously caught exception.  See L<perlguts/"Exception Handling">.
 #endif
 
 #define XS_APIVERSION_BOOTCHECK						\
-    Perl_xs_apiversion_bootcheck(aTHX_ ST(0), STR_WITH_LEN("v" PERL_API_VERSION_STRING))
+    Perl_xs_apiversion_bootcheck(ST(0), STR_WITH_LEN("v" PERL_API_VERSION_STRING))
 
 #ifdef NO_XSLOCKS
 #  define dXCPT             dJMPENV; int rEtV = 0
diff --git a/embed.fnc b/embed.fnc
index c79f767..1a6d24f 100644
--- a/embed.fnc
+++ b/embed.fnc
@@ -2698,7 +2698,7 @@ Xpo	|void	|xs_version_bootcheck|U32 items|U32 ax|NN const char *xs_p \
 				|STRLEN xs_len
 : This function is an implementation detail. The public API for this is
 : XS_APIVERSION_BOOTCHECK
-Xpo	|void	|xs_apiversion_bootcheck|NN SV *module|NN const char *api_p \
+Xpon	|void	|xs_apiversion_bootcheck|NN SV *module|NN const char *api_p \
 				|STRLEN api_len
 
 #ifndef HAS_STRLCAT
diff --git a/pod/perlclib.pod b/pod/perlclib.pod
index 7f86f1b..03dce25 100644
--- a/pod/perlclib.pod
+++ b/pod/perlclib.pod
@@ -119,6 +119,9 @@ There is no equivalent to C<fgets>; one should use C<sv_gets> instead:
                                               / strGT(s1,s2)
  strncmp(s1, s2, n)             strnNE(s1, s2, n) / strnEQ(s1, s2, n)
 
+ memcmp(p1, p2, n)              memNE(p1, p2, n)
+ !memcmp(p1, p2, n)             memEQ(p1, p2, n)
+
 Notice the different order of arguments to C<Copy> and C<Move> than used
 in C<memcpy> and C<memmove>.
 
diff --git a/pod/perldiag.pod b/pod/perldiag.pod
index 262d04b..df15a17 100644
--- a/pod/perldiag.pod
+++ b/pod/perldiag.pod
@@ -4349,6 +4349,11 @@ is equivalent to v5.100.
 recent than the currently running version.  How long has it been since
 you upgraded, anyway?  See L<perlfunc/require>.
 
+=item Perl API version %s of %s does not match %s
+
+(F) The XS module in question was complied against a different incompatible
+version of Perl than the one that has loaded the XS module.
+
 =item PERL_SH_DIR too long
 
 (F) An error peculiar to OS/2.  PERL_SH_DIR is the directory to find the
diff --git a/proto.h b/proto.h
index 197566a..72d1470 100644
--- a/proto.h
+++ b/proto.h
@@ -5168,9 +5168,9 @@ PERL_CALLCONV void	Perl_write_to_stderr(pTHX_ SV* msv)
 #define PERL_ARGS_ASSERT_WRITE_TO_STDERR	\
 	assert(msv)
 
-PERL_CALLCONV void	Perl_xs_apiversion_bootcheck(pTHX_ SV *module, const char *api_p, STRLEN api_len)
-			__attribute__nonnull__(pTHX_1)
-			__attribute__nonnull__(pTHX_2);
+PERL_CALLCONV void	Perl_xs_apiversion_bootcheck(SV *module, const char *api_p, STRLEN api_len)
+			__attribute__nonnull__(1)
+			__attribute__nonnull__(2);
 #define PERL_ARGS_ASSERT_XS_APIVERSION_BOOTCHECK	\
 	assert(module); assert(api_p)
 
diff --git a/util.c b/util.c
index 2391c62..489890d 100644
--- a/util.c
+++ b/util.c
@@ -5391,34 +5391,16 @@ Perl_xs_version_bootcheck(pTHX_ U32 items, U32 ax, const char *xs_p,
 }
 
 void
-Perl_xs_apiversion_bootcheck(pTHX_ SV *module, const char *api_p,
+Perl_xs_apiversion_bootcheck(SV *module, const char *api_p,
 			     STRLEN api_len)
 {
-    SV *xpt = NULL;
-    SV *compver = Perl_newSVpvn_flags(aTHX_ api_p, api_len, SVs_TEMP);
-    SV *runver;
-
     PERL_ARGS_ASSERT_XS_APIVERSION_BOOTCHECK;
 
-    /* This might croak  */
-    compver = upg_version(compver, 0);
-    /* This should never croak */
-    runver = new_version(PL_apiversion);
-    if (vcmp(compver, runver)) {
-	SV *compver_string = vstringify(compver);
-	SV *runver_string = vstringify(runver);
-	xpt = Perl_newSVpvf(aTHX_ "Perl API version %"SVf
-			    " of %"SVf" does not match %"SVf,
-			    SVfARG(compver_string), SVfARG(module),
-			    SVfARG(runver_string));
-	Perl_sv_2mortal(aTHX_ xpt);
-
-	SvREFCNT_dec(compver_string);
-	SvREFCNT_dec(runver_string);
-    }
-    SvREFCNT_dec(runver);
-    if (xpt)
-	Perl_croak_sv(aTHX_ xpt);
+    if(api_len != sizeof("v" PERL_API_VERSION_STRING)-1
+	|| memNE(api_p, "v" PERL_API_VERSION_STRING, sizeof("v" PERL_API_VERSION_STRING)-1)) {
+	Perl_croak_nocontext("Perl API version %s of %"SVf" does not match %s",
+			    api_p, SVfARG(module), "v" PERL_API_VERSION_STRING);
+    }
 }
 
 /*
-- 
1.7.9.msysgit.0


```
</details>

**Comment by p5pRT** [comment]
### From @cpansprout

On Thu Nov 06 20&#8203;:58&#8203;:55 2014\, bulk88 wrote&#8203;:
> This is a bug report for perl from bulk88@&#8203;hotmail\.com\,
> generated with the help of perlbug 1\.40 running under perl 5\.21\.4\.
> 
> 
> \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
> \[Please describe your issue here\]
> 
> See attached patch\.

Thank you for spotting such a blatant instance of overengineering that most of us failed to notice\!  I have applied it as 0e42d607f\.

\-\- 

Father Chrysostomos

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
@cpansprout - Status changed from 'open' to 'resolved'

**Comment by p5pRT** [comment]
### From @mauke

Am Do 06\. Nov 2014\, 21&#8203;:52&#8203;:05\, sprout schrieb&#8203;:
> On Thu Nov 06 20&#8203;:58&#8203;:55 2014\, bulk88 wrote&#8203;:
> > This is a bug report for perl from bulk88@&#8203;hotmail\.com\,
> > generated with the help of perlbug 1\.40 running under perl 5\.21\.4\.
> >
> >
> > \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
> > \[Please describe your issue here\]
> >
> > See attached patch\.
> 
> Thank you for spotting such a blatant instance of overengineering that
> most of us failed to notice\!  I have applied it as 0e42d607f\.

\+\(F\) The XS module in question was complied against a different incompatible

s/complied/compiled/

**Comment by p5pRT** [comment]
### From @cpansprout

On Thu Nov 06 21&#8203;:56&#8203;:04 2014\, mauke\- wrote&#8203;:
> Am Do 06\. Nov 2014\, 21&#8203;:52&#8203;:05\, sprout schrieb&#8203;:
> > On Thu Nov 06 20&#8203;:58&#8203;:55 2014\, bulk88 wrote&#8203;:
> > > This is a bug report for perl from bulk88@&#8203;hotmail\.com\,
> > > generated with the help of perlbug 1\.40 running under perl 5\.21\.4\.
> > >
> > >
> > > \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
> > > \[Please describe your issue here\]
> > >
> > > See attached patch\.
> > 
> > Thank you for spotting such a blatant instance of overengineering that
> > most of us failed to notice\!  I have applied it as 0e42d607f\.
> 
> \+\(F\) The XS module in question was complied against a different incompatible
> 
> s/complied/compiled/

Thank you\.  I have fixed it in d792985a\.

Why didn’t a spelling checker catch that? :\-\)

\-\- 

Father Chrysostomos
